### PR TITLE
Incremental unwinding of one specified loop -- rebase

### PR DESCRIPTION
**Pull request by xbauch**
Incremental unwinding and assertion checking of a specified loop.

This is a rebase of #4361.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by peterschrammel — APPROVED** [review]

**Comment by hannes-steffenhagen-diffblue** [comment]
@tautschnig Were your comments from https://github.com/diffblue/cbmc/pull/4361 addressed?

File: src/goto-symex/symex_target_equation.cpp [hunk 1]
```diff
@@ -318,6 +318,19 @@ void symex_target_equationt::constraint(
   merge_ireps(SSA_step);
 }
 
+void symex_target_equationt::convert_without_assertions(
```
Comment (thk123) [inline src/goto-symex/symex_target_equation.cpp:321]: > Could this be refactored so that we don't have almost-the-same-code twice?

https://github.com/diffblue/cbmc/pull/4361#discussion_r292643973

Looks to be outstanding

File: src/goto-symex/symex_target_equation.cpp [hunk 2]
```diff
@@ -458,13 +472,13 @@ void symex_target_equationt::convert_assertions(
...
 
-  if(number_of_assertions==1)
+  if(number_of_assertions == 1 && optimized_for_single_assertions)
   {
     for(auto &step : SSA_steps)
     {
       // ignore already converted assertions in the error trace
       if(step.is_assert() && step.converted)
-        step.ignore = true;
+        step.hidden = true;
```
Comment (thk123) [inline src/goto-symex/symex_target_equation.cpp:474]: > That makes the comment above seem rather odd?

https://github.com/diffblue/cbmc/pull/4361#discussion_r292549222

Outstanding

File: src/goto-symex/symex_target_equation.cpp [hunk 3]
```diff
@@ -491,7 +505,7 @@ void symex_target_equationt::convert_assertions(
   {
     // ignore already converted assertions in the error trace
     if(step.is_assert() && step.converted)
-      step.ignore = true;
+      step.hidden = true;
```
Comment (thk123) [inline src/goto-symex/symex_target_equation.cpp:501]: Ditto

File: src/goto-symex/goto_symex.h [hunk 4]
```diff
@@ -164,6 +164,17 @@ class goto_symext
   /// symbolic execution from that state.
   bool should_pause_symex;
 
+  /// If this flag is set to true then assertions will be temporarily ignored
+  /// by the symbolic executions.
+  bool ignore_assertions = false;
+
+  virtual bool check_break(
```
Comment (thk123) [inline src/goto-symex/goto_symex.h]: > Documentation?

https://github.com/diffblue/cbmc/pull/4361#discussion_r292644774

outstanding

File: regression/cbmc-incr-oneloop/Makefile [hunk 5]
```diff
@@ -1,10 +1,10 @@
 default: tests.log
 
 test:
-	@../test.pl -c "perl -e 'alarm shift @ARGV; exec @ARGV' 30 ../../../src/cbmc/cbmc --slice-formula"
+	@../test.pl -c "perl -e 'alarm shift @ARGV; exec @ARGV' 8 ../../../src/cbmc/cbmc --slice-formula"
 
 tests.log: ../test.pl
-	@../test.pl -c "perl -e 'alarm shift @ARGV; exec @ARGV' 30 ../../../src/cbmc/cbmc --slice-formula"
+	@../test.pl -c "perl -e 'alarm shift @ARGV; exec @ARGV' 8 ../../../src/cbmc/cbmc --slice-formula"
```
Comment (thk123) [inline regression/cbmc-incr-oneloop/Makefile:8]: Michael would like a comment here explaining the perl invocation is for the timeout

File: src/goto-checker/symex_bmc_incremental_one_loop.cpp [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    incr_loop_id(options.get_option("incremental-loop")),
+    incr_max_unwind(
+      options.is_set("unwind-max") ? options.get_signed_int_option("unwind-max")
+                                   : std::numeric_limits<unsigned>::max()),
+    incr_min_unwind(
+      options.is_set("unwind-min") ? options.get_signed_int_option("unwind-min")
+                                   : 0)
+{
+  ignore_assertions =
+    options.get_bool_option("ignore-properties-before-unwind-min");
```
Comment (thk123) [inline src/goto-checker/symex_bmc_incremental_one_loop.cpp:40]: Perhaps a question for @peterschrammel 

> When was this option introduced and where is it documented?

https://github.com/diffblue/cbmc/pull/4361#discussion_r292647453

**Review by thk123 — COMMENTED** [review]
Comments from @tautschnig on original PR that are still outstanding

File: src/goto-checker/symex_bmc_incremental_one_loop.cpp (hunk 6 above)
Comment (peterschrammel) [reply src/goto-checker/symex_bmc_incremental_one_loop.cpp:40]: It has been introduced as part of this PR. Maybe it hasn't been documented as part of thIs PR, though.

Comment (thk123) [reply src/goto-checker/symex_bmc_incremental_one_loop.cpp:40]: Hmm, seems to be documented here @tautschnig : https://github.com/diffblue/cbmc/pull/5217/files#diff-e58a25459f0e8ca5a6c351b1d9d75167R225

**Comment by thk123** [comment]
Niavely looking at coverage we should add the following tests:
 - a pretty shortest trace (e.g. with `--beautify` turned on for one of the tests)
 - building a non-shortest trace for a property
 - building a full trace

(dunno how you exercise those last two options) 

Also, we should add a trivial test for  checking exit with non-zero exit code if ran with --paths as well as --incremental-loop just to stop any boneheaded mistakes in future 

Also, less nit picky, but it seems like `symex_bmc_incremental_one_loopt` is not tested with any `loop_unwind_handlers`. This seems like it should be added

File: src/goto-symex/symex_target_equation.cpp (hunk 1 above)
Comment (xbauch) [reply src/goto-symex/symex_target_equation.cpp:321]: Refactored.

File: src/goto-symex/symex_target_equation.cpp (hunk 2 above)
Comment (xbauch) [reply src/goto-symex/symex_target_equation.cpp:474]: Comment changed (also below).

File: src/goto-symex/goto_symex.h (hunk 4 above)
Comment (xbauch) [reply src/goto-symex/goto_symex.h]: Documented (and refactored a bit).

File: regression/cbmc-incr-oneloop/Makefile (hunk 5 above)
Comment (xbauch) [reply regression/cbmc-incr-oneloop/Makefile:8]: Comment added.

File: src/goto-symex/goto_symex.h [hunk 7]
```diff
@@ -164,6 +164,22 @@ class goto_symext
...
+  /// \brief Defines condition for interrupting symbolic execution for a
+  ///   specific loop
+  ///
+  /// False constant by default, over-ridden by incremental BMC implementation
+  ///   to allow breaking if the unwinding the user specified loop
+  /// \param id: the loop identifier
+  /// \param unwind: current unwinding counter
+  /// \return true if the symbolic execution is to be interrupted for checking
+  virtual bool check_break(
+    const irep_idt &id,
```
Comment (thk123) [inline src/goto-symex/goto_symex.h]: :pick: is this the loop id? given how frequent `id`s are used in this code base, I'd probably call it `loop_id`

**Review by thk123 — APPROVED** [review]
AFAICT all of @tautschnig comments from #4361 have been addressed.

**Comment by xbauch** [comment]
> * a pretty shortest trace (e.g. with `--beautify` turned on for one of the tests)

It seems that the refinement solver does not support beatification

**Comment by xbauch** [comment]
> * a pretty shortest trace (e.g. with `--beautify` turned on for one of the tests)
> * building a non-shortest trace for a property
> * building a full trace

I have a suspicion that some of the `build_*trace` functions are only there to override the default and cannot really be called in any of our factory-produced verifiers. I'll keep digging.

File: src/goto-symex/goto_symex.h (hunk 7 above)
Comment (xbauch) [reply src/goto-symex/goto_symex.h]: Done.

**Comment by xbauch** [comment]
@peterschrammel In an attempt to improve coverage, the last commit adds a test for trace production. I've also suggested commenting out the implementation of the "beautified" trace and the "full" trace since I don't think they can be exercised by any of the current verifier combinations. Is it change acceptable?

File: src/goto-checker/single_loop_incremental_symex_checker.cpp [hunk 8]
```diff
@@ -168,25 +168,33 @@ operator()(propertiest &properties)
...
-  goto_tracet goto_trace;
-  build_goto_trace(
-    equation,
-    equation.SSA_steps.end(),
-    property_decider.get_decision_procedure(),
-    ns,
-    goto_trace);
-
-  return goto_trace;
+  UNREACHABLE;
```
Comment (peterschrammel) [inline src/goto-checker/single_loop_incremental_symex_checker.cpp]: 🚫 Write a test for it if you want to cover it.

**Review by peterschrammel — CHANGES_REQUESTED** [review]

File: src/goto-checker/single_loop_incremental_symex_checker.cpp (hunk 8 above)
Comment (xbauch) [reply src/goto-checker/single_loop_incremental_symex_checker.cpp]: Done.

File: src/goto-checker/symex_bmc_incremental_one_loop.cpp (hunk 6 above)
Comment (xbauch) [reply src/goto-checker/symex_bmc_incremental_one_loop.cpp:40]:
```suggestion
    incr_min_unwind >= 2 && options.get_bool_option("ignore-properties-before-unwind-min");
```
@peterschrammel or at least this way it was in 5.4.

Comment (peterschrammel) [reply src/goto-checker/symex_bmc_incremental_one_loop.cpp:40]: Any observable effect of adding this?

Comment (xbauch) [reply src/goto-checker/symex_bmc_incremental_one_loop.cpp:40]: Well since the test(s) in question do not set the `unwind-min` I assume it is 1 by default. So the ignoring would not even occur. So yes, the test passes, but rather trivially.

File: regression/cbmc-incr-oneloop/ignore-before-unwind/main.c [hunk 9]
```diff
@@ -0,0 +1,14 @@
+int main()
```
Comment (thk123) [inline regression/cbmc-incr-oneloop/ignore-before-unwind/main.c:1]: :pick: typo in commit message `KNOWBUG` -> `KNOWNBUG`

File: regression/cbmc-incr-oneloop/alarm2/fancy_trace_test.desc [hunk 10]
```diff
@@ -0,0 +1,12 @@
+KNOWNBUG
+main.c
+--incremental-loop main.0 --unwind-min 5 --unwind-max 10 --trace --beautify --no-sat-preprocessor --stop-on-fail
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
+--
+^warning: ignoring
+--
+This will fail the invariant on `build_goto_trace.cpp:320` asserting certain
```
Comment (thk123) [inline regression/cbmc-incr-oneloop/alarm2/fancy_trace_test.desc:10]: Could you raise a bug on cbmc and put the issue number here.

File: regression/cbmc-incr-oneloop/alarm2/trace_test.desc [hunk 11]
```diff
@@ -0,0 +1,9 @@
+CORE
+main.c
+--incremental-loop main.0 --unwind-min 5 --unwind-max 10 --trace
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
+^State \d+ file main.c function havocIO line 501 thread \d+$
+--
+^warning: ignoring
```
Comment (thk123) [inline regression/cbmc-incr-oneloop/alarm2/trace_test.desc:9]: A description of why you added this test (e.g. tests the trace production for ??)

**Comment by thk123** [comment]
Maybe you already knew this, but the key flag is ` --ignore-properties-before-unwind-min`, disabling that causes it to find the failure

File: src/goto-checker/symex_bmc_incremental_one_loop.cpp (hunk 6 above)
Comment (xbauch) [reply src/goto-checker/symex_bmc_incremental_one_loop.cpp:40]: On the other hand, the test is still wrong if you set the `unwind-min` (to 10 for example) even with the suggested change.

File: regression/cbmc-incr-oneloop/ignore-before-unwind/main.c [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+{
+  __CPROVER_bool assumption = (__CPROVER_bool)0;
+  __CPROVER_bool property = (__CPROVER_bool)1;
+
+  while(1)
+  {
+    __CPROVER_assume(property == 1);
```
Comment (peterschrammel) [inline regression/cbmc-incr-oneloop/ignore-before-unwind/main.c:7]: What is the purpose of the unused `assumption` variable?

Comment (xbauch) [reply regression/cbmc-incr-oneloop/ignore-before-unwind/main.c:7]: Hard to say, but if you comment it out, it suddenly starts to work.

Comment (peterschrammel) [reply regression/cbmc-incr-oneloop/ignore-before-unwind/main.c:7]: I mean the assumption variable, not the assumption statement.

Comment (xbauch) [reply regression/cbmc-incr-oneloop/ignore-before-unwind/main.c:7]: Sorry, yes, that one is irrelevant.

**Comment by xbauch** [comment]
@peterschrammel I've added the check for when to start ignoring properties (eebfcd98675f6392b6ec962481f5465c543ddb40) and cleaned up the test.

**Review by martin-cs — APPROVED** [review]
If this branch has the history I think it does it might be a candidate for the longest running branch as it predates even VSD!

**Comment by thk123** [comment]
Just noting the travis job has passed, but for some reason the icon has not refreshed on here

File: regression/cbmc-incr-oneloop/unwindset-more-loops1/test.desc [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
+^Unwinding loop main\.0 iteration 5
+^Not unwinding loop main\.1 iteration 2
+^Not unwinding loop main\.2 iteration 8
+^Unwinding loop main\.3 iteration 3
+--
+^warning: ignoring
+^Unwinding loop main\.0 iteration 6
```
Comment (peterschrammel) [inline regression/cbmc-incr-oneloop/unwindset-more-loops1/test.desc]: Missing new line at end of file

**Review by peterschrammel — APPROVED** [review]

File: regression/cbmc-incr-oneloop/unwindset-more-loops1/test.desc (hunk 13 above)
Comment (xbauch) [reply regression/cbmc-incr-oneloop/unwindset-more-loops1/test.desc]: Fixed (I had to rerun the CI anyway).
